### PR TITLE
add cpp implementation to usaco-743

### PR DESCRIPTION
**Pull request by WilliamYuan42**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.

### CODE REVIEW

File: solutions/gold/usaco-743.mdx [hunk 1]
```diff
@@ -76,5 +75,63 @@ public class ModernArt2 {
...
 ```
+
 </JavaSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+const int mx = 1e5 + 5;
+int col[mx], start[mx], fin[mx], n, ans = 0;
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-743.mdx:86]: declarations on separate lines

File: solutions/gold/usaco-743.mdx [hunk 2]
```diff
@@ -76,5 +75,63 @@ public class ModernArt2 {
...
 }
 ```
+
 </JavaSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+const int mx = 1e5 + 5;
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-743.mdx:85]: is the +5 needed

File: solutions/gold/usaco-743.mdx [hunk 3]
```diff
@@ -76,5 +75,63 @@ public class ModernArt2 {
...
+int col[mx], start[mx], fin[mx], n, ans = 0;
+
+int main() {
+	ios::sync_with_stdio(false);
+	cin.tie(nullptr);
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-743.mdx:90]: remove if runs w/o

File: solutions/gold/usaco-743.mdx [hunk 4]
```diff
@@ -76,5 +75,63 @@ public class ModernArt2 {
...
+	freopen("art2.in", "r", stdin);
+	freopen("art2.out", "w", stdout);
+	fill(start, start + mx, 1 << 30);
+	fill(fin, fin + mx, -(1 << 30));
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-743.mdx:94]: lmao why not just use vectors

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/gold/usaco-743.mdx (hunk 4 above)
Comment (WilliamYuan42) [reply solutions/gold/usaco-743.mdx:94]: idk lol

File: solutions/gold/usaco-743.mdx [hunk 5]
```diff
@@ -76,5 +75,61 @@ public class ModernArt2 {
...
+using namespace std;
+
+int main() {
+	freopen("art2.in", "r", stdin);
+	freopen("art2.out", "w", stdout);
+	int n;
+	cin >> n;
+	vector<int> col(n + 1);
+	vector<int> start(n + 1, 1 << 30);
+	vector<int> fin(n + 1, -(1 << 30));
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-743.mdx:93]: use INT32_MAX and INT32_MIN instead of these weird numbers

File: solutions/gold/usaco-743.mdx [hunk 6]
```diff
@@ -76,5 +75,61 @@ public class ModernArt2 {
...
+
+	// get start, end points
+	for (int i = 1; i <= n; i++) {
+		cin >> col[i];
+		start[col[i]] = min(start[col[i]], i);
+		fin[col[i]] = max(fin[col[i]], i);
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-743.mdx:99]: maybe `end` instead of `fin`

File: solutions/gold/usaco-743.mdx [hunk 7]
```diff
@@ -76,5 +75,61 @@ public class ModernArt2 {
...
+	}
+
+	start[0] = 0, fin[0] = n + 1;
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-743.mdx:102]: separate lines

File: solutions/gold/usaco-743.mdx [hunk 8]
```diff
@@ -76,5 +75,61 @@ public class ModernArt2 {
...
+		}
+
+		// check for partial intersections
+		// for example, in painting [1, 1, 2, 1, 2]
+		// color 1 has start, end at 1, 4
+		// color 2 has start, end at 3, 5
+		// it is impossible to interweave them
+		// since both colors are used only once
+		// to paint a segment of the painting
+		// thus, the painting is invalid
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-743.mdx:119]:
```suggestion
```
should be explained in the editorial

Comment (SansPapyrus683) [reply solutions/gold/usaco-743.mdx:119]: or copy the comment from the java code, yeah you should do that instead

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/gold/usaco-743.mdx [hunk 9]
```diff
@@ -76,5 +75,61 @@ public class ModernArt2 {
...
+		cin >> col[i];
+		start[col[i]] = min(start[col[i]], i);
+		end[col[i]] = max(end[col[i]], i);
+	}
+
+	int ans = 0;
+	start[0] = 0;
+	end[0] = n + 1;
+	stack<int> active;  // keep track of segments
+
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-743.mdx:106]:
```suggestion
```

**Review by SansPapyrus683 — APPROVED** [review]
good enough

**Comment by SansPapyrus683** [comment]
no, i'm just very picky about this kinda stuff
another pm probably would've had it merged in 10 minutes

File: solutions/gold/usaco-743.mdx [hunk 10]
```diff
@@ -76,5 +75,60 @@ public class ModernArt2 {
...
+			cout << "-1\n";
+			return 0;
+		}
+
+		// remove color at end
+		if (i == end[col[i]]) active.pop();
+	}
+
+	// '0' is not a color.
+	cout << ans - 1 << "\n";
```
Comment (ChuangSheep) [inline solutions/gold/usaco-743.mdx:129]:
```suggestion
	cout << ans - 1 << endl;
```

File: solutions/gold/usaco-743.mdx [hunk 11]
```diff
@@ -76,5 +75,60 @@ public class ModernArt2 {
...
+			ans = max(ans, (int)active.size());
+		}
+
+		/*
+		 * Invalid because there are 2 partially intersecting segments.
+		 * For instance if we have a finished painting of 1 1 2 1 2,
+		 * it is impossible if we can only paint 1 once.
+		 */
+		if (cur != active.top()) {
+			cout << "-1\n";
```
Comment (ChuangSheep) [inline solutions/gold/usaco-743.mdx:120]:
```suggestion
			cout << "-1" << endl;
```

**Review by ChuangSheep — APPROVED** [review]
nice

**Comment by ChuangSheep** [comment]
@SansPapyrus683 merge?

**Pull request merged**